### PR TITLE
Fix an issue `tsh` uses wrong default username for auto-user provisioning enabled databases in remote clusters

### LINKED ISSUE #35901: `tsh db connect` to leaf database with auto-user provisioning uses wrong username by default
Expected behavior:
```bash
$ tsh db connect --cluster leaf.example.com --db-name sales my-db-with-auto-user`
```
Should succeed without providing `--db-user`.

Current behavior:
```bash
$ tsh db connect --cluster leaf.example.com --db-name sales my-db-with-auto-user`
psql: error: connection to server at "localhost" (::1), port 50764 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
connection to server at "localhost" (127.0.0.1), port 50764 failed: please use your Teleport username ("remote-<username>-leaf.example.com") to connect instead of "<username>"
```

Workaround:
```bash
$ tsh db connect --cluster leaf.example.com --db-name sales my-db-with-auto-user` --db-user "remote-<username>-leaf.example.com"
```

Bug details:
- Teleport version: v14
- Recreation steps: as above

When accessing remote clusters, the Username is mapped as follows
https://github.com/gravitational/teleport/blob/57a9ec4b2114cebc95dde4cee9be34f57f3a9933/lib/authz/permissions.go#L531-L536

### PR DESCRIPTION
**Pull request by greedy52**
Fixes #35901:
- #35901 

changelog: Fix an issue `tsh` uses wrong default username for auto-user provisioning enabled databases in remote clusters

Changes:
- Fixed "Allowed Users" in `tsh db ls` for remote auto-user database
- Fixed the default user when `--db-user` is not specified
- Improved the engine message when `--db-user` doesn't match
- Improved `tsh proxy db` output to include database user and database name

Examples:
```bash
 $ tsh db ls --cluster $leafcluster --search mariadb
Name                          Description Allowed Users                                                      Labels                 Connect 
----------------------------- ----------- ------------------------------------------------------------------ ---------------------- ------- 
self-hosted-mariadb                       [*]                                                                                               
self-hosted-mariadb-auto-drop             [remote-root-teleport.root.dev.aws.stevexin.me] (Auto-provisioned) db-auto-user-mode=drop         
self-hosted-mariadb-auto-keep             [remote-root-teleport.root.dev.aws.stevexin.me] (Auto-provisioned) db-auto-user-mode=keep         

$ tsh proxy db --cluster $leafcluster  self-hosted-mariadb-auto-drop 
Started DB proxy on 127.0.0.1:49215
To avoid port randomization, you can choose the listening port using the --port flag.

Teleport Connect is a desktop app that can manage database proxies for you.
Learn more at https://goteleport.com/docs/connect-your-client/teleport-connect/#connecting-to-a-database

Use the following credentials to connect to the self-hosted-mariadb proxy:
  ca_file=/Users/stevehuang/.tsh-root/keys/teleport.root.dev.aws.stevexin.me/cas/teleport.root.dev.aws.stevexin.me.pem
  cert_file=/Users/stevehuang/.tsh-root/keys/teleport.root.dev.aws.stevexin.me/root-db/teleport.dev.aws.stevexin.me/self-hosted-mariadb-auto-drop-x509.pem
  key_file=/Users/stevehuang/.tsh-root/keys/teleport.root.dev.aws.stevexin.me/root

Your database user is "remote-root-teleport.root.dev.aws.stevexin.me".

$ tsh db connect --db-user aaa --db-name test self-hosted-mariadb-auto-drop --cluster $leafcluster
ERROR 1105 (HY000): please use your mapped remote username ("remote-root-teleport.root.dev.aws.stevexin.me") to connect instead of "aaa"
ERROR: exit status 1
```

### CODE REVIEW

File: tool/tsh/common/db.go [hunk 1]
```diff
@@ -887,18 +887,6 @@ func (d *databaseInfo) checkAndSetDefaults(cf *CLIConf, tc *client.TeleportClien
...
-	// username was provided explicitly.
-	if needDBUser && db.GetAdminUser().Name != "" {
-		log.Debugf("Defaulting to Teleport username %q as database username.", tc.Username)
-		d.Username = tc.Username
-		needDBUser = false
-	}
-	if !needDBUser && !needDBName {
-		return nil
-	}
-
```
Comment (greedy52) [inline tool/tsh/common/db.go:901]: removing this and the logic will fall below to call `getDefaultDBUser`

File: lib/srv/db/common/session.go [hunk 2]
```diff
@@ -94,3 +95,27 @@ func (c *Session) WithUserAndDatabase(user string, defaultDatabase string) *Sess
...
+		return nil
+	}
+
+	if c.DatabaseUser == c.Identity.Username {
+		return nil
+	}
+
+	if c.AuthContext != nil && authz.IsRemoteUser(*c.AuthContext) {
+		return trace.AccessDenied("please use your mapped remote username (%q) to connect instead of %q",
+			c.Identity.Username, c.DatabaseUser)
```
Comment (smallinsky) [inline lib/srv/db/common/session.go:116]: Nit: 
Could we add a short docs troubleshooting update and explain what this error means. 

User needs to be aware of remote clusters setup and know that username on remote cluster is mapped to `remote-main-cluster-username` without this this massage is quite vague. 


We can just add a troubleshooting error "please use your mapped remote username" and explain this error.

**Review by smallinsky — APPROVED** [review]

File: lib/srv/db/common/session.go (hunk 2 above)
Comment (greedy52) [reply lib/srv/db/common/session.go:116]: we added a note on the long name here: https://github.com/gravitational/teleport/pull/36648/files#diff-026e51d24f3932c0b0dca15ed5e7c96362d9217a96311bc3c69cff193f4e2356

but a troubleshooting will definitely help when that error pops up. will do separately. thanks!

**Comment by GavinFrazar** [comment]
LGTM and I tested local/remote cluster connection to mongodb w/ auto user provisioning

**Review by GavinFrazar — APPROVED** [review]

**Pull request merged**